File: src/python/multiprocessing/process_pool.py
"""
Process Pool Example

This script demonstrates using ProcessPoolExecutor to efficiently manage
a pool of worker processes for parallel execution of CPU-bound tasks.

Key Concepts:
- ProcessPoolExecutor manages process lifecycle automatically
- Processes are reused across tasks (avoids spawn overhead)
- submit() returns a Future representing the pending result
- as_completed() yields Futures as tasks finish (not submission order)

When to Use:
- CPU-bound tasks (computation, data processing, image manipulation)
- Tasks that can be parallelized across multiple cores
- When you need true parallelism (bypassing Python's GIL)

For I/O-bound tasks, consider ThreadPoolExecutor instead, as it has
lower overhead and the GIL doesn't affect I/O operations.
"""

import concurrent.futures
import random
import time


def worker(task_id):
    """
    Worker function that simulates a CPU-bound task.

    Args:
        task_id: Identifier for the task

    Returns:
        Tuple of (task_id, result)
    """
    print(f"Task {task_id} is starting...")
    time.sleep(random.uniform(1, 3))  # Simulate CPU-bound work
    result = task_id * 2
    print(f"Task {task_id} is finished. Result: {result}")
    return task_id, result


def main():
    num_tasks = 10
    num_workers = 3

    with concurrent.futures.ProcessPoolExecutor(max_workers=num_workers) as executor:
        # Submit all tasks to the executor
        futures = [executor.submit(worker, i) for i in range(num_tasks)]

        # Process results as they complete (not in submission order)
        for future in concurrent.futures.as_completed(futures):
            task_id, result = future.result()
            print(f"Task {task_id} result collected: {result}")


if __name__ == "__main__":
    main()

File: src/cpp/multiprocessing/process_pool.cpp
/**
 * Process Pool
 *
 * This example demonstrates a simple process pool - a fixed number
 * of worker processes that execute tasks from a queue.
 *
 * Key concepts:
 * - Limits the number of concurrent processes to avoid resource exhaustion
 * - Reuses process slots as tasks complete
 * - Parent process manages the pool and collects results via exit codes
 * - Child exit status is retrieved via WEXITSTATUS macro
 *
 * Process pools are useful when:
 * - Tasks are CPU-bound and benefit from parallel execution
 * - You want to limit system resource usage
 * - Each task needs memory isolation
 */

#include <algorithm>
#include <chrono>
#include <iostream>
#include <sys/wait.h>
#include <thread>
#include <unistd.h>
#include <vector>

class ProcessPool {
public:
    explicit ProcessPool(size_t max_workers)
        : max_workers_(max_workers) {}

    void submit(int task_id) {
        // If pool is full, wait for a worker to finish
        if (workers_.size() >= max_workers_) {
            wait_for_one();
        }

        pid_t pid = fork();
        if (pid < 0) {
            std::cerr << "fork() failed\n";
            return;
        }
        if (pid == 0) {
            // Child: execute task and exit with result
            execute_task(task_id);
            _exit(task_id * 2);  // Return result via exit code
        } else {
            workers_.push_back(pid);
        }
    }

    void wait_all() {
        while (!workers_.empty()) {
            wait_for_one();
        }
    }

private:
    void wait_for_one() {
        int status;
        pid_t finished = wait(&status);

        if (WIFEXITED(status)) {
            std::cout << "Task result: " << WEXITSTATUS(status) << "\n";
        }

        auto it = std::find(workers_.begin(), workers_.end(), finished);
        if (it != workers_.end()) {
            workers_.erase(it);
        }
    }

    void execute_task(int task_id) {
        std::cout << "Task " << task_id << " starting (PID: " 
                  << getpid() << ")\n";
        std::this_thread::sleep_for(std::chrono::seconds(1));
        std::cout << "Task " << task_id << " finished\n";
    }

    size_t max_workers_;
    std::vector<pid_t> workers_;
};

int main() {
    constexpr int num_tasks = 10;
    constexpr int pool_size = 3;

    ProcessPool pool(pool_size);

    for (int i = 0; i < num_tasks; ++i) {
        pool.submit(i);
    }

    pool.wait_all();
    std::cout << "All tasks completed\n";

    return 0;
}
